### PR TITLE
apps/example/timer : Add FRT measurement before and after the timer e…

### PR DESCRIPTION
**Pull request by jeongchanKim**
…vent

Add FRT measurement for summarizing the elpased time.
Expected result is like below:
>> Summary
        Expected 500000, Estimated 500259, Ratio 0.000517964
        PASS : Estimated Elapsed time is within the expected range.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [ae836232dd13bbb99f7ffb89061afdfc60fd2aa6] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ae836232dd13bbb99f7ffb89061afdfc60fd2aa6] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [adff57dc926fb49e0bda46e4edfaca2620bd0569] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [adff57dc926fb49e0bda46e4edfaca2620bd0569] - Code Rule Check (C++) OK.

File: apps/examples/timer/timer_main.c [hunk 1]
```diff
@@ -146,12 +155,41 @@ static pthread_addr_t timer_thread(pthread_addr_t arg)
...
+	/* Open the FRTimer device */
+	if (pargs->devno == 0) {
+		frt_dev = 1;
+	} else {
+		frt_dev = 0;
+	}
+	sprintf(path, TIMER_DEVNAME, frt_dev);
+	fprintf(stdout, "FRT Open %s\n", path);
+
+	frt_fd = open(path, O_RDONLY);
```
Comment (YiHyunJin) [inline apps/examples/timer/timer_main.c]: I thinks that frt_fd should close.

**Comment by seinfra** [comment]
Target : [52b3fb87041696be1da6093c4aafac1b744f4601] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [52b3fb87041696be1da6093c4aafac1b744f4601] - Code Rule Check (C++) OK.

File: apps/examples/timer/Kconfig [hunk 2]
```diff
@@ -11,3 +11,12 @@ config EXAMPLES_TIMER
 		This application is an example program that opens /dev/timer0
 		to set the timer and registers the callback which is called at
 		the end of the timer.
+
+if EXAMPLES_TIMER
+config FRT_MEASUREMENT
```
Comment (sunghan-chang) [inline apps/examples/timer/Kconfig]: Let's have prefix, 'EXAMPLES_TIMER`

**Comment by seinfra** [comment]
Target : [e821a7431465c4389906d1fa14d87b5c83cb1c6e] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [e821a7431465c4389906d1fa14d87b5c83cb1c6e] - Code Rule Check (C++) OK.

File: apps/examples/timer/timer_main.c (hunk 1 above)
Comment (jeongchanKim) [reply apps/examples/timer/timer_main.c]: Done.

File: apps/examples/timer/timer_main.c [hunk 3]
```diff
@@ -146,12 +159,46 @@ static pthread_addr_t timer_thread(pthread_addr_t arg)
...
 	}
 
+#ifdef CONFIG_EXAMPLES_TIMER_FRT_MEASUREMENT
+	/* Open the FRTimer device */
+	if (pargs->devno == 0) {
+		frt_dev = 1;
+	} else {
+		frt_dev = 0;
+	}
+	sprintf(path, TIMER_DEVNAME, frt_dev);
```
Comment (manoharahk) [inline apps/examples/timer/timer_main.c]: Better to use snprintf

File: apps/examples/timer/timer_main.c (hunk 1 above)
Comment (manoharahk) [reply apps/examples/timer/timer_main.c]: Better to validate frt_fd before it gets used with ioctl

**Comment by seinfra** [comment]
Target : [6148fe00a4b21773751e7ebea15d49bca6b6f322] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [6148fe00a4b21773751e7ebea15d49bca6b6f322] - Code Rule Check (C++) OK.

File: apps/examples/timer/timer_main.c (hunk 1 above)
Comment (jeongchanKim) [reply apps/examples/timer/timer_main.c]: Done. Thanks.

File: apps/examples/timer/timer_main.c [hunk 4]
```diff
@@ -146,12 +159,50 @@ static pthread_addr_t timer_thread(pthread_addr_t arg)
...
+	}
+	snprintf(path, _POSIX_PATH_MAX, TIMER_DEVNAME, frt_dev);
+	fprintf(stdout, "FRT Open %s\n", path);
+
+	frt_fd = open(path, O_RDONLY);
+	if (frt_fd < 0) {
+		fprintf(stderr, "ERROR: Failed to open Free Run Timer: %d\n", errno);
+		goto error;
+	}
+	ioctl(frt_fd, TCIOC_SETFREERUN, TRUE);
```
Comment (sijupunnnoose) [inline apps/examples/timer/timer_main.c]: what if the ioctl fails?

Comment (jeongchanKim) [reply apps/examples/timer/timer_main.c]: It is just set the register value. ioctl always returns OK.

**Comment by seinfra** [comment]
Target : [b0161b20adfb38775076555ef937b8febae90bd1] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [b0161b20adfb38775076555ef937b8febae90bd1] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [f0f074f55706fcc61d1f6da14f2480626d20bdae] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [f0f074f55706fcc61d1f6da14f2480626d20bdae] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f1711d7e763b6a2f04978691e49f4d013dc7e739] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [f1711d7e763b6a2f04978691e49f4d013dc7e739] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [bb787c9934f41989a7f4b962f2d5e038a3b6f21f] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [bb787c9934f41989a7f4b962f2d5e038a3b6f21f] - Code Rule Check (C++) OK.

File: apps/examples/timer/timer_main.c [hunk 5]
```diff
@@ -146,12 +161,62 @@ static pthread_addr_t timer_thread(pthread_addr_t arg)
...
+		fprintf(stderr, "ERROR: Failed to start Free Run Timer: %d\n", errno);
+		goto error;
+	}
+
+	if (ioctl(frt_fd, TCIOC_GETSTATUS, (unsigned long)(uintptr_t)&before) < 0) {
+		fprintf(stderr, "ERROR: Failed to get Free Run Timer status: %d\n", errno);
+		goto error;
+	}
+
+	expected_time = count * intval;
```
Comment (davidfather) [inline apps/examples/timer/timer_main.c]: Please, move this line above, i.e., before timer setting.

File: apps/examples/timer/timer_main.c [hunk 6]
```diff
@@ -146,12 +161,62 @@ static pthread_addr_t timer_thread(pthread_addr_t arg)
...
-			fprintf(stdout, "time limits(%d)\n", ++nbr);
+		if (sigwaitinfo(&sig_set, NULL) == EXAMPLE_TIMER_SIGNO);
 	}
+#ifdef CONFIG_EXAMPLES_TIMER_FRT_MEASUREMENT
+	if (ioctl(frt_fd, TCIOC_GETSTATUS, (unsigned long)(uintptr_t)&after) < 0) {
+		fprintf(stderr, "ERROR: Failed to get Free Run Timer status: %d\n", errno);
+	}
+	time_diff = after.timeleft - before.timeleft;
+#endif
+
```
Comment (davidfather) [inline apps/examples/timer/timer_main.c]: Please, move "ioctl(fd, TCIOC_STOP, 0)" here.

File: apps/examples/timer/timer_main.c [hunk 7]
```diff
@@ -146,12 +161,62 @@ static pthread_addr_t timer_thread(pthread_addr_t arg)
...
 	/* Stop the timer */
 	fprintf(stdout, "Stop the timer\n");
+#ifdef CONFIG_EXAMPLES_TIMER_FRT_MEASUREMENT
+	fprintf(stdout, ">> Summary\n");
```
Comment (davidfather) [inline apps/examples/timer/timer_main.c]: Please, move "ioctl(frt_fd, TCIOC_STOP, 0)" here.

**Comment by seinfra** [comment]
Target : [ff3862720d699c4b3ef8330a7989c3c9841378d7] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [ff3862720d699c4b3ef8330a7989c3c9841378d7] - Code Rule Check (C++) OK.

File: apps/examples/timer/timer_main.c (hunk 6 above)
Comment (jeongchanKim) [reply apps/examples/timer/timer_main.c]: Done.

File: apps/examples/timer/timer_main.c (hunk 7 above)
Comment (jeongchanKim) [reply apps/examples/timer/timer_main.c]: Done.

**Comment by seinfra** [comment]
Target : [3bdafe0befc195b5ea883553c17896b9c71f9301] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [3bdafe0befc195b5ea883553c17896b9c71f9301] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [8729fff7ee92f2f3a89372ec51894f497f84a7fa] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [8729fff7ee92f2f3a89372ec51894f497f84a7fa] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ffff7bd2aba81d8c63c5e81152059d45b32e4e32] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ffff7bd2aba81d8c63c5e81152059d45b32e4e32] - Code Rule Check OK.

File: apps/examples/timer/timer_main.c [hunk 8]
```diff
@@ -146,24 +161,76 @@ static pthread_addr_t timer_thread(pthread_addr_t arg)
...
+		fprintf(stderr, "ERROR: Failed to stop the Free Run Timer: %d\n", errno);
+	}
+#endif
+	fprintf(stdout, "Stop the timer\n");
+#ifdef CONFIG_EXAMPLES_TIMER_FRT_MEASUREMENT
+	fprintf(stdout, ">> Summary\n");
+
+	time_diff = after.timeleft - before.timeleft;
+	fprintf(stdout, "\tExpected %d, Estimated %d, Ratio %.9f\n", expected_time, time_diff, ((float)time_diff / (float)expected_time) - 1.0);
+	if (((float)expected_time * (1 + ALLOWABLE_PERCENTAGE / 100)) > (float)time_diff) {
```
Comment (hasw7569) [inline apps/examples/timer/timer_main.c]: please, casting float (1 + ALLOWABLE_PERCENTAGE / 100).

**Comment by seinfra** [comment]
Target : [f5301af0a2b4b7e950073e321a261a9c680ee6b2] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [f5301af0a2b4b7e950073e321a261a9c680ee6b2] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [d5c5715676cc130dab96bc15c2182102721d1e2b] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [d5c5715676cc130dab96bc15c2182102721d1e2b] - Code Rule Check (C++) OK.

File: apps/examples/timer/timer_main.c [hunk 9]
```diff
@@ -146,24 +161,76 @@ static pthread_addr_t timer_thread(pthread_addr_t arg)
...
+	}
+
+	if (ioctl(frt_fd, TCIOC_GETSTATUS, (unsigned long)(uintptr_t)&before) < 0) {
+		fprintf(stderr, "ERROR: Failed to get Free Run Timer status: %d\n", errno);
+		goto error;
+	}
+
+#endif
+	while (count--) {
+		if (sigwaitinfo(&sig_set, NULL) == EXAMPLE_TIMER_SIGNO);
```
Comment (jeongarmy) [inline apps/examples/timer/timer_main.c]: I think `if` comparison is not needed.

**Comment by seinfra** [comment]
Target : [0cf020d456cb40b09d9e66822d43e520938d5f5f] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [0cf020d456cb40b09d9e66822d43e520938d5f5f] - Code Rule Check OK.

**Pull request merged**